### PR TITLE
Add per build id timestamp for when it was last made set default

### PR DESCRIPTION
**Pull request by bergundy**
This is a continuation of the work done in #4524.

We now store a timestamp per build id denoting when it was last made default in its set.

This information will later be used by the scavenger to ensure that we don't delete build ids that have recently been set default giving enough time for outstanding tasks to be completed and visibility index to be updated.

### CODE REVIEW

File: service/matching/version_sets_test.go [hunk 1]
```diff
@@ -123,20 +122,17 @@ func TestNewDefaultUpdate(t *testing.T) {
...
-				BuildIds:                    []*persistencespb.BuildId{{Id: "1", State: persistencespb.STATE_ACTIVE, StateUpdateTimestamp: &clock}},
-				DefaultUpdateTimestamp:      &clock,
+				BuildIds:                    []*persistencespb.BuildId{{Id: "1", State: persistencespb.STATE_ACTIVE, StateUpdateTimestamp: &clock, SetDefaultUpdateTimestamp: &clock}},
 				QueueDefaultUpdateTimestamp: &clock,
 			},
 			{
 				SetIds:                      []string{hashBuildId("2")},
-				BuildIds:                    []*persistencespb.BuildId{{Id: "2", State: persistencespb.STATE_ACTIVE, StateUpdateTimestamp: &nextClock}},
-				DefaultUpdateTimestamp:      &nextClock,
+				BuildIds:                    []*persistencespb.BuildId{{Id: "2", State: persistencespb.STATE_ACTIVE, StateUpdateTimestamp: &nextClock, SetDefaultUpdateTimestamp: &nextClock}},
```
Comment (Sushisource) [inline service/matching/version_sets_test.go:135]: Re-using `mkBuildId` would shorten things considerably

Comment (Sushisource) [reply service/matching/version_sets_test.go:135]: Same for the rest of the other tests

File: service/matching/version_sets_test.go [hunk 2]
```diff
@@ -863,12 +838,11 @@ func TestMergeSets(t *testing.T) {
 	bothSetIds := mergeSetIDs([]string{hashBuildId("1")}, []string{hashBuildId("2")})
 	assert.Equal(t, bothSetIds, updatedData.GetVersionSets()[1].GetSetIds())
 	assert.Equal(t, initialData.GetVersionSets()[2].QueueDefaultUpdateTimestamp, updatedData.GetVersionSets()[1].QueueDefaultUpdateTimestamp)
-	assert.Equal(t, nextClock, *updatedData.GetVersionSets()[1].DefaultUpdateTimestamp)
+	assert.Equal(t, nextClock, *updatedData.VersionSets[1].BuildIds[len(updatedData.VersionSets[1].BuildIds)-1].SetDefaultUpdateTimestamp)
```
Comment (Sushisource) [inline service/matching/version_sets_test.go:841]: This access is a mouthful. Maybe make a little helper for these 3 spots

File: proto/internal/temporal/server/api/persistence/v1/task_queues.proto [hunk 3]
```diff
@@ -39,6 +39,10 @@ message BuildId {
     // (-- api-linter: core::0142::time-field-type=disabled
     //     aip.dev/not-precedent: Using HLC instead of wall clock. --)
     temporal.server.api.clock.v1.HybridLogicalClock state_update_timestamp = 3;
+    // HLC timestamp representing when this build id was last made default in its version set.
```
Comment (Sushisource) [inline proto/internal/temporal/server/api/persistence/v1/task_queues.proto:42]:
```suggestion
    // HLC timestamp representing when this build id was last made default in its version set, if ever.
```

File: proto/internal/temporal/server/api/persistence/v1/task_queues.proto [hunk 4]
```diff
@@ -39,6 +39,10 @@ message BuildId {
...
+    // (-- api-linter: core::0142::time-field-type=disabled
+    //     aip.dev/not-precedent: Using HLC instead of wall clock. --)
+    temporal.server.api.clock.v1.HybridLogicalClock set_default_update_timestamp = 4;
```
Comment (Sushisource) [inline proto/internal/temporal/server/api/persistence/v1/task_queues.proto:45]: I think `became_default_timestamp` might be an easier to understand name

**Review by Sushisource — APPROVED** [review]
Nothing blocking

File: proto/internal/temporal/server/api/persistence/v1/task_queues.proto (hunk 3 above)
Comment (bergundy) [reply proto/internal/temporal/server/api/persistence/v1/task_queues.proto:42]: We currently don't expose an API to add a build id without making it set default in a set

File: proto/internal/temporal/server/api/persistence/v1/task_queues.proto (hunk 4 above)
Comment (bergundy) [reply proto/internal/temporal/server/api/persistence/v1/task_queues.proto:45]: Fair

**Review by yiminc — APPROVED** [review]

File: proto/internal/temporal/server/api/persistence/v1/task_queues.proto (hunk 3 above)
Comment (Sushisource) [reply proto/internal/temporal/server/api/persistence/v1/task_queues.proto:42]: Ah, good point

**Pull request merged**